DID Algo Method

bcessa opened the pull request:
The `did:algo` method provides a Decentralized Identifier implementation build on top of the Algorand Blockchain and IPFS storage.

The open source tools are available (pre-build) on:
https://github.com/algorandfoundation/did-algo/releases

For more information:
https://github.com/algorandfoundation/did-algo

kdenhartog commented on packages/did-core-test-server/suites/implementations/did-algo.json:
```diff
@@ -0,0 +1,86 @@
+            "id": "did:algo:f30304e2-e702-4f47-b242-2e9e064f6eef#master",
+            "type": "Ed25519VerificationKey2018",
+            "controller": "did:algo:f30304e2-e702-4f47-b242-2e9e064f6eef",
+            "publicKeyMultibase": "zCh9PDTZzeWxk2WdH4M1e8k2951D5D11jz7Uti9HRBGiK"
```
Heads up, your public keys may be dropped by a json-ld document loader here because the contexts don't define these properties.

I'd suggest relying on the v3-unstable context to avoid this.

msporny replied:
err, I don't think v3-unstable is going to survive. I suggest using the pattern as shown in the latest did:key method: https://github.com/w3c/did-test-suite/blob/main/packages/did-core-test-server/suites/implementations/did-key-2020-db.json#L14-L18

kdenhartog replied:
I couldn't find the proper suit that gave Ed25519VerificationKey2018 and publicKeyMultibase other than v3-unstable. However, I think your recommendation of upgrading to the 2020 suite is actually the better solution here.

bcessa replied:
We have updated the contexts used to:

```json
"https://w3id.org/security/suites/ed25519-2020/v1",
"https://w3id.org/security/suites/x25519-2020/v1"
```

kdenhartog commented on packages/did-core-test-server/suites/implementations/did-algo.json:
```diff
@@ -0,0 +1,86 @@
+            "id": "did:algo:f30304e2-e702-4f47-b242-2e9e064f6eef#algo-connect",
+            "type": "did.algorand.foundation.ExternalService",
+            "serviceEndpoint": "https://did.algorand.foundation",
+            "extensions": [
```
same issue here - this property will get dropped by a json-ld document loader

bcessa replied:
Addressed and validated with external tools here: https://github.com/w3c/did-test-suite/pull/203#issuecomment-964417077

kdenhartog commented:
While the entry here is technically correct, this should be considered as an additional data point that JSON-LD representation is going to likely present issues with any relying party who is depending on the json-ld representation because the production of good json-ld representations is hard.

msporny commented:
> While the entry here is technically correct

Any entry that contains an `@context` and doesn't round-trip through JSON-LD is breaking a SHOULD requirement in the spec. It's not illegal, but it's not something a method should do w/o a very good reason.

> JSON-LD representation is going to likely present issues with any relying party who is depending on the JSON-LD representation because the production of good JSON-LD representations is hard.

We should add tooling to the DID Test Suite that takes a document through JSON-LD expansion/compaction and checks to see if properties are dropped. That will make DID Method implementers that use the test suite aware of these issues early in the process.

kdenhartog commented:
> Any entry that contains an `@context` and doesn't round-trip through JSON-LD is breaking a SHOULD requirement in the spec. It's not illegal, but it's not something a method should do w/o a very good reason.

Yup, that was my understanding as well so i didn't want to request changes but did want to call it out so that more people are aware of these things.

>We should add tooling to the DID Test Suite that takes a document through JSON-LD expansion/compaction and checks to see if properties are dropped. That will make DID Method implementers that use the test suite aware of these issues early in the process.

Agreed, I can file an issue for this so that there's a simpler way to discover this going forward but probably won't get around to writing the test for a bit.

bcessa commented:
Hi, thanks for the feedback. We have addressed the comments above.

A basic document generated by the method is of the form:

```json
{
  "@context": [
    "https://www.w3.org/ns/did/v1",
    "https://w3id.org/security/v1",
    "https://w3id.org/security/suites/ed25519-2020/v1",
    "https://w3id.org/security/suites/x25519-2020/v1",
    "https://did-ns.aidtech.network/v1"
  ],
  "id": "did:algo:56da1708-eead-4e2d-9558-f53d684003fd",
  "verificationMethod": [
    {
      "id": "did:algo:56da1708-eead-4e2d-9558-f53d684003fd#master",
      "type": "Ed25519VerificationKey2020",
      "controller": "did:algo:56da1708-eead-4e2d-9558-f53d684003fd",
      "publicKeyMultibase": "z8sdBpaoefmAi7iFbY96ZkP1ff9VoRSRTJs4NQEyLfunH"
    }
  ],
  "authentication": [
    "did:algo:56da1708-eead-4e2d-9558-f53d684003fd#master"
  ],
  "service": [
    {
      "id": "did:algo:56da1708-eead-4e2d-9558-f53d684003fd#algo-connect",
      "type": "did.algorand.foundation.ExternalService",
      "serviceEndpoint": "https://did.algorand.foundation",
      "extensions": [
        {
          "id": "algo-address",
          "version": "0.1.0",
          "data": [
            {
              "address": "GZAURFQR7VRUL2TRVMUJ3G3IGFYSFXBWHBFFQBEITXWDLFVVHTICFNY2LE",
              "asset": "ALGO",
              "network": "testnet"
            }
          ]
        }
      ]
    }
  ]
}
```

We have validated the JSON-LD expansion/compaction operations to make sure there are no properties being dropped.

Additionally we used some third-party tools available:
- [JSON-LD Checker](https://jsonld-checker.com/?json=%257B%250A%2520%2520%2522%2540context%2522%253A%2520%255B%250A%2520%2520%2520%2520%2522https%253A%252F%252Fwww.w3.org%252Fns%252Fdid%252Fv1%2522%252C%250A%2520%2520%2520%2520%2522https%253A%252F%252Fw3id.org%252Fsecurity%252Fv1%2522%252C%250A%2520%2520%2520%2520%2522https%253A%252F%252Fw3id.org%252Fsecurity%252Fsuites%252Fed25519-2020%252Fv1%2522%252C%250A%2520%2520%2520%2520%2522https%253A%252F%252Fw3id.org%252Fsecurity%252Fsuites%252Fx25519-2020%252Fv1%2522%252C%250A%2520%2520%2520%2520%2522https%253A%252F%252Fdid-ns.aidtech.network%252Fv1%2522%250A%2520%2520%255D%252C%250A%2520%2520%2522id%2522%253A%2520%2522did%253Aalgo%253A56da1708-eead-4e2d-9558-f53d684003fd%2522%252C%250A%2520%2520%2522verificationMethod%2522%253A%2520%255B%250A%2520%2520%2520%2520%257B%250A%2520%2520%2520%2520%2520%2520%2522id%2522%253A%2520%2522did%253Aalgo%253A56da1708-eead-4e2d-9558-f53d684003fd%2523master%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522type%2522%253A%2520%2522Ed25519VerificationKey2020%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522controller%2522%253A%2520%2522did%253Aalgo%253A56da1708-eead-4e2d-9558-f53d684003fd%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522publicKeyMultibase%2522%253A%2520%2522z8sdBpaoefmAi7iFbY96ZkP1ff9VoRSRTJs4NQEyLfunH%2522%250A%2520%2520%2520%2520%257D%250A%2520%2520%255D%252C%250A%2520%2520%2522authentication%2522%253A%2520%255B%250A%2520%2520%2520%2520%2522did%253Aalgo%253A56da1708-eead-4e2d-9558-f53d684003fd%2523master%2522%250A%2520%2520%255D%252C%250A%2520%2520%2522service%2522%253A%2520%255B%250A%2520%2520%2520%2520%257B%250A%2520%2520%2520%2520%2520%2520%2522id%2522%253A%2520%2522did%253Aalgo%253A56da1708-eead-4e2d-9558-f53d684003fd%2523algo-connect%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522type%2522%253A%2520%2522did.algorand.foundation.ExternalService%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522serviceEndpoint%2522%253A%2520%2522https%253A%252F%252Fdid.algorand.foundation%2522%252C%250A%2520%2520%2520%2520%2520%2520%2522extensions%2522%253A%2520%255B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522id%2522%253A%2520%2522algo-address%2522%252C%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522version%2522%253A%2520%25220.1.0%2522%252C%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522data%2522%253A%2520%255B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522address%2522%253A%2520%2522GZAURFQR7VRUL2TRVMUJ3G3IGFYSFXBWHBFFQBEITXWDLFVVHTICFNY2LE%2522%252C%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522asset%2522%253A%2520%2522ALGO%2522%252C%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2522network%2522%253A%2520%2522testnet%2522%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%255D%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%255D%250A%2520%2520%2520%2520%257D%250A%2520%2520%255D%250A%257D%250A)
- [JSON-LD Playground](https://json-ld.org/playground/#json-ld=%7B%0A%20%20%22%40context%22%3A%20%5B%0A%20%20%20%20%22https%3A%2F%2Fwww.w3.org%2Fns%2Fdid%2Fv1%22%2C%0A%20%20%20%20%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fv1%22%2C%0A%20%20%20%20%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fsuites%2Fed25519-2020%2Fv1%22%2C%0A%20%20%20%20%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fsuites%2Fx25519-2020%2Fv1%22%2C%0A%20%20%20%20%22https%3A%2F%2Fdid-ns.aidtech.network%2Fv1%22%0A%20%20%5D%2C%0A%20%20%22id%22%3A%20%22did%3Aalgo%3A56da1708-eead-4e2d-9558-f53d684003fd%22%2C%0A%20%20%22verificationMethod%22%3A%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%22id%22%3A%20%22did%3Aalgo%3A56da1708-eead-4e2d-9558-f53d684003fd%23master%22%2C%0A%20%20%20%20%20%20%22type%22%3A%20%22Ed25519VerificationKey2020%22%2C%0A%20%20%20%20%20%20%22controller%22%3A%20%22did%3Aalgo%3A56da1708-eead-4e2d-9558-f53d684003fd%22%2C%0A%20%20%20%20%20%20%22publicKeyMultibase%22%3A%20%22z8sdBpaoefmAi7iFbY96ZkP1ff9VoRSRTJs4NQEyLfunH%22%0A%20%20%20%20%7D%0A%20%20%5D%2C%0A%20%20%22authentication%22%3A%20%5B%0A%20%20%20%20%22did%3Aalgo%3A56da1708-eead-4e2d-9558-f53d684003fd%23master%22%0A%20%20%5D%2C%0A%20%20%22service%22%3A%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%22id%22%3A%20%22did%3Aalgo%3A56da1708-eead-4e2d-9558-f53d684003fd%23algo-connect%22%2C%0A%20%20%20%20%20%20%22type%22%3A%20%22did.algorand.foundation.ExternalService%22%2C%0A%20%20%20%20%20%20%22serviceEndpoint%22%3A%20%22https%3A%2F%2Fdid.algorand.foundation%22%2C%0A%20%20%20%20%20%20%22extensions%22%3A%20%5B%0A%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%22id%22%3A%20%22algo-address%22%2C%0A%20%20%20%20%20%20%20%20%20%20%22version%22%3A%20%220.1.0%22%2C%0A%20%20%20%20%20%20%20%20%20%20%22data%22%3A%20%5B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22address%22%3A%20%22GZAURFQR7VRUL2TRVMUJ3G3IGFYSFXBWHBFFQBEITXWDLFVVHTICFNY2LE%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22asset%22%3A%20%22ALGO%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22network%22%3A%20%22testnet%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%5D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%7D%0A%20%20%5D%0A%7D%0A)

msporny commented:
@bcessa wrote:
> ```json
>     "https://w3id.org/security/v1",
> ```

I think you /should/ be able to get away with dropping the v1 security context. We're in the process of breaking out all the cryptosuites into their own JSON-LD contexts, which you are using in `https://w3id.org/security/suites/ed25519-2020/v1` and `https://w3id.org/security/suites/x25519-2020/v1`.

You might also consider changing this:

```
 "type": "did.algorand.foundation.ExternalService"
```

to this:

```
 "type": "AlgorandExternalService"
```

and then scoping your `extensions` property to any object that is of type `AlgorandExternalService`. That should protect your `extensions` term from conflicting with w/ any future `extensions` property that someone else might do.

Again, just suggestions, you don't need to do those things for this PR to go in.

Do you want us to wait until you make those changes, or merge?

bcessa commented:
Thanks for your feedback @msporny 👍🏼 

I'll go ahead and remove the old security context and adjust the service type to `AlgorandExternalService` right now before merging if that's ok.

We'll also adjust the extension scope and add a few more attributes in the coming days, hopefully this won't be a blocker for the merge, is that correct?

msporny commented:
> We'll also adjust the extension scope and add a few more attributes in the coming days, hopefully this won't be a blocker for the merge, is that correct?

Correct, not a blocker for the merge. You can update your implementation at any time.

I see that you've applied https://github.com/w3c/did-test-suite/pull/203/commits/486114760013ef201d81cad381c7291f8a3fe8fa, do you want us to merge now?

bcessa commented:
> I see that you've applied [4861147](https://github.com/w3c/did-test-suite/commit/486114760013ef201d81cad381c7291f8a3fe8fa), do you want us to merge now?

Sure, if all comments are now considered addressed that would be great. Thanks.
